proposals/readonly-struct-init.md

```diff
@@ -0,0 +1,161 @@
+a simple field initialization when used as an object initializer. 
+
+``` csharp 
+struct Name {
+    // Backing field: public readonly string <FirstName>__BackingField;
+    public string FirstName { get; }
+    // Backing field: public readonly string <LastName>__BackingField;
+    public string LastNmae { get; }
+}
+
+var name = new Name { FirstName = "Jared", LastName = "Parsons" };
+
+// Emitted as 
+var name = new Name { <FirstName>__BackingField = "Jared", <LastName>__BackingField = "Parsons" };
+```
+
+This does mean that the feature will only apply to auto-implemented properties that are 
+re-compiled with a new compiler. The compiler today emits the fields as `private readonly` and 
+hence will be inaccessible to calling code. Upon recompile the backing fields will be emitted 
+as specified above.
+
+### Verification
+
+As is the case with several features added since C# 7.2 this feature will not pass any known 
+peverify implementation as it makes no real distinction between read only and mutable memory. 
+However it is still desirable for C#, and other languages, to emit IL that can be verified 
+with appropriate adjustments to the verification spec. Eventually allowing for projects like
+[ILVerify](https://github.com/dotnet/corert/tree/master/src/ILVerify) to take the place of 
+peverify today. 
+
+This rules could be adjusted to allow for mutation of a `readonly` field following an `.initobj` 
+or `.call .ctor` instruction if:
+
+1. The reciever of the call is a local 
+1. The type of the reciever is a `struct` or `ref struct`
+1. There are no `ref` locals which refer to that local
+
+The `readonly` fields would remain mutable until one of the following occured:
+
+1. The reciever was passed as any `ref` to another function except when being used as `this`
+1. A `ref` local was taken to the local or any of its members
+
+This may force conforming compilers to create extra locals in some cases to allow for a
+successful verification of the emitted code.
+
+## Considerations
+
+### Type Parameters
+
+The feature does not apply to type parameters constrained to a `struct`. There is just no case 
+where such a type would have a field (`readonly` or not) or an auto-implemented property. In the
+future if we allow for `struct` inheritance it's possible this could change and the decision 
+re-evaluated.
+
+## classes
+
+This can't apply to `class` instances as it's much easier to smuggle a reference out of the 
+constructor. The `this` value can be easily stored in a `static` making it accessible to the 
+calling code long before the object initialization has completed. 
+
+``` csharp
+static C Global;
+
+class C { 
+    public readonly int Field;
+
+    public C() { 
+        Global = this;
+    }
+}
+
+int SomeMethod() => Global.Field++;
+
+// The instance is observed by caller while mutations are still in progress.
+var c = new C() { Field = SomeMethod()}
```

Comment:
Can you please explain why this is a problem? It's not clear to me. So `SomeMethod` would be able to observe a different value of a readonly field, but that is already possible if we call `SomeMethod` from inside the constructor. Why is this form worse? We're essentially making it as if the object initializer was part of the constructor. That means it can mutate the fields and also observe the mutations.

Reply:
> Why is this form worse? 

It's not worse. It's just demonstrating that you can observe the value changing after the constructor has completed. The cases you speak of apply from within the constructor only (and methods transitively called from it).